File: java/src/solutions/2840_check_if_strings_can_be_made_equal_with_operations_II.java
class Solution {
  public boolean checkStrings(String s1, String s2) {
    if (s1.length() != s2.length())
      return false;

    int[] map = new int[26];
    for (int i = 0; i < s1.length(); i += 2) {
      map[s1.charAt(i) - 'a']++;
      map[s2.charAt(i) - 'a']--;
    }
    for (int x : map) {
      if (x != 0)
        return false;
    }

    for (int i = 1; i < s1.length(); i += 2) {
      map[s1.charAt(i) - 'a']++;
      map[s2.charAt(i) - 'a']--;
    }
    for (int x : map) {
      if (x != 0)
        return false;
    }
    return true;
  }

  void main() {
    assert checkStrings("abcd", "cdab");
    assert checkStrings("abcdba", "cabdab");
    assert !checkStrings("abcd", "dacb");
    assert !checkStrings("abe", "bea");
  }
}

File: rust/src/bin/2840_check_if_strings_can_be_made_equal_with_operations_II.rs
struct Solution {}
impl Solution {
    pub fn check_strings(s1: String, s2: String) -> bool {
        if s1.len() != s2.len() {
            return false;
        }

        let mut count = [0; 26];
        for i in (0..s1.len()).step_by(2) {
            count[(s1.as_bytes()[i] - b'a') as usize] += 1;
            count[(s2.as_bytes()[i] - b'a') as usize] -= 1;
        }
        if count.iter().any(|&n| n != 0) {
            return false;
        }

        for i in (1..s1.len()).step_by(2) {
            count[(s1.as_bytes()[i] - b'a') as usize] += 1;
            count[(s2.as_bytes()[i] - b'a') as usize] -= 1;
        }

        count.iter().all(|&n| n == 0)
    }
}

fn main() {
    assert!(Solution::check_strings(
        "abcd".to_string(),
        "cdab".to_string()
    ));
    assert!(Solution::check_strings(
        "abcdba".to_string(),
        "cabdab".to_string()
    ));
    assert!(!Solution::check_strings(
        "abcd".to_string(),
        "dacb".to_string()
    ));
    assert!(!Solution::check_strings(
        "abe".to_string(),
        "bea".to_string()
    ));
}
